### PR TITLE
Add configurations for userns

### PR DESCRIPTION
**Pull request by avisiedo**
- (removed) Add configuration for updating runc rpm package.
- (removed) Add configuration for updating cri-o rpm package.
- Add configuration for updating rpm packages.
- Add configuration for subids.
- Add a userns composition for creating all the configs
  with one command.
- Update configuration for cgroup-v2.

Changes based on [Demo: namespaced systemd workloads on OpenShift](https://frasertweedale.github.io/blog-redhat/posts/2021-07-22-openshift-systemd-workload-demo.html).

Thanks to @frasertweedale

### CODE REVIEW

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        # $ oc adm node-logs node/NODE -u install-cri-o
+        contents: |
+          [Unit]
+          Description=Install custom cri-o
+          # https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
+          After=network-online.target
+
+          [Service]
+          Type=oneshot
+          ExecStart=bash -c 'rpm -q "$( rpm -q ${RPM_CRI_O} )" || rpm-ostree override replace --reboot "${RPM_CRI_O}"'
```
Comment (frasertweedale) [inline config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: Side note: we are usually overriding `runc`, not `cri-o` :)

From this command construction, I assume that we were unable to get it to work with only
`rpm-ostree override replace --idempotent --reboot`.  That is, was it unconditionally rebooting?
And is that why we must explicitly test the package version via `rpm` program?

I don't like that the package must be unconditionally downloaded upon every reboot, in order to perform
this test.  Perhaps two separate variables are better - one for the package name and the other for the
URL?  The `Makefile` can even take care of this, deriving the package name from the RPM and setting a
variable - rather than this being done by systemd on the worker node, every time it boots.

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      units:
+      - name: "install-cri-o.service"
+        enabled: true
+        # For getting the logs run:
+        # $ oc adm node-logs node/NODE -u install-cri-o
+        contents: |
+          [Unit]
+          Description=Install custom cri-o
+          # https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
+          After=network-online.target
```
Comment (frasertweedale) [inline config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: This unit does not work for me on OCP 4.10.  Logs:
```
% oc adm node-logs node/ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx -u rpm-overrides.service                                                                    
-- Logs begin at Thu 2021-12-02 02:52:45 UTC, end at Thu 2021-12-02 05:20:26 UTC. --
-- No entries --
-- Logs begin at Thu 2021-12-02 02:52:45 UTC, end at Thu 2021-12-02 05:20:26 UTC. --
Dec 02 05:00:24.321949 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: Starting Install RPM overrides...
Dec 02 05:00:24.339044 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: package runc-1.0.2-992.rhaos4.10.el8.x86_64 is not installed
Dec 02 05:00:25.177411 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Downloading 'https://ftweedal.fedorapeople.org/runc-1.0.2-992.rhaos4.10.el8.x86_64.rpm'... done!
Dec 02 05:00:29.025484 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Checking out tree d95b9a4...done
Dec 02 05:00:29.089007 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: No enabled rpm-md repositories.
Dec 02 05:00:29.203517 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Importing rpm-md...done
Dec 02 05:00:29.212301 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Resolving dependencies...done
Dec 02 05:00:29.214332 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Applying 1 override
Dec 02 05:00:29.306524 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Processing packages...done
Dec 02 05:00:29.309166 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running pre scripts...done
Dec 02 05:00:29.310469 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running post scripts...done
Dec 02 05:00:30.232707 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Running posttrans scripts...done
Dec 02 05:00:30.420456 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Writing rpmdb...done
Dec 02 05:00:32.387732 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Writing OSTree commit...done
Dec 02 05:00:35.136481 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: Staging deployment...done
Dec 02 05:00:35.137815 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx sh[1355]: error: Bus owner changed, aborting. This likely means the daemon crashed; check logs with `journalctl -xe`.
Dec 02 05:00:35.141900 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: rpm-overrides.service: Main process exited, code=exited, status=1/FAILURE
Dec 02 05:00:35.142056 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: rpm-overrides.service: Failed with result 'exit-code'.
Dec 02 05:00:35.142501 ci-ln-m7jtknk-72292-9xv9c-worker-c-v74jx systemd[1]: Failed to start Install RPM overrides.
```

Perhaps it is being executed too early in boot sequence?  Perhaps it needs to run `After=rpm-ostreed.service`?

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 1 above)
Comment (avisiedo) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: about cri-o: Yea, I know! but the effort was a little to get the config for cri-o too, so we have the automation just in case.
about the command: I was using previously`'rpm-ostree install --idempotent --reboot`, but I should have tested wrongly; when I was doing final tests I saw that it was not installing the package and I searched for a workaround.

Thanks @frasertweedale for point out the uncondicional download, I will update the PR for using two variables, so it is downloaded only when it is needed.

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 2 above)
Comment (avisiedo) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: Thanks @frasertweedale ; I tested on OCP 4.9.5; I am installing an OCP 4.10 for fixing this.

Comment (avisiedo) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: I have tried in OCP 4.9.9 and it was successful too; the output I get is the below:

```raw
-- Logs begin at Fri 2021-11-26 04:40:34 UTC, end at Thu 2021-12-02 21:08:58 UTC. --
Dec 02 21:07:01 crc-x4qnm-master-0 bash[1663]: package runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64 is not installed
Dec 02 21:07:01 crc-x4qnm-master-0 systemd[1]: Starting Install custom runc...
Dec 02 21:07:05 crc-x4qnm-master-0 bash[1663]: Downloading 'https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm'... done!
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Checking out tree b690184...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: No enabled rpm-md repositories.
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Importing rpm-md...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Resolving dependencies...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Applying 1 override and 5 overlays
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Processing packages...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Running pre scripts...done
Dec 02 21:07:07 crc-x4qnm-master-0 bash[1663]: Running post scripts...done
Dec 02 21:07:08 crc-x4qnm-master-0 bash[1663]: Running posttrans scripts...done
Dec 02 21:07:08 crc-x4qnm-master-0 bash[1663]: Writing rpmdb...done
Dec 02 21:07:09 crc-x4qnm-master-0 bash[1663]: Writing OSTree commit...done
Dec 02 21:07:10 crc-x4qnm-master-0 bash[1663]: Staging deployment...done
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: install-runc.service: Succeeded.
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: Stopped Install custom runc.
Dec 02 21:07:11 crc-x4qnm-master-0 systemd[1]: install-runc.service: Consumed 161ms CPU time
-- Reboot --
Dec 02 21:08:00 crc-x4qnm-master-0 bash[1648]: runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: Starting Install custom runc...
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: install-runc.service: Succeeded.
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: Started Install custom runc.
Dec 02 21:08:00 crc-x4qnm-master-0 systemd[1]: install-runc.service: Consumed 15ms CPU time
```
Comparing with the output that you get, I can see an error message at the end; the message indicate the daemon crashed; I want to try in OCP 4.10 to dig more on this and replay it. I am having some situations to get the cluster ready.

By the way, your traces inspired me for getting a more generalized overlay for installing the rpm packages, thank you; I have updated the PR so that now we have only one overlay that can be customized for the POOL and the list of packages. I have tested in CodeReadyContainers so far.

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 1 above)
Comment (frasertweedale) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: I think the new approach is good, avoiding the unconditional download during bootup and only downloading it if the desired RPM is not installed. :+1:

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 2 above)
Comment (frasertweedale) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: Unfortunately, I get the exact same failure on OCP 4.9 (nightly) via cluster-bot.  Must be something different about how nodes are managed in CRC.

Recommend you install cluster(s) on PSI or via cluster-bot for testing.

File: config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst [hunk 3]
```diff
@@ -20,11 +20,15 @@ spec:
           [Unit]
           Description=Install custom runc
           # https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/
-          After=network-online.target rpm-ostreed.service
+          After=network-online.target
+          # After=network-online.target
+          StartLimitIntervalSec=10
+          StartLimitBurst=30
```
Comment (frasertweedale) [inline config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst]: I don't think these are really required... or if we do decide we want it (e.g. prevent unlimited download of external file due to unlimited restart), then these values are not going to achieve the desired effect.

From `systemd.unit(5)`, these settings work together to inhibit further restarts if the service restarts *burst* times in *interval-sec* seconds.  But here we are downloading an external RPM, so it seems unlikely that we would ever hit the rate of 30 in 10 seconds (or 3 per second).

So I would suggest either remove the setting (it is for a hack after all, not for producion use), or else change these settings to "10 times in 5 minutes" or something like that.  Alternatively, under `[Service]` specify a high `RestartSec` to configure a longer delay between restart attempts (perhaps 1 minute).

File: config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst [hunk 4]
```diff
@@ -20,11 +20,15 @@ spec:
...
 
           [Service]
-          Type=oneshot
+          Type=simple
```
Comment (frasertweedale) [inline config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst]: [nit] we can remove this line; `simple` is the default when specifying `ExecStart`.

**Review by frasertweedale — DISMISSED** [review]
It works, good stuff.  A couple of nits (see comments).  Feel free to amend, or merge it as is.

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 2 above)
Comment (avisiedo) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: Thanks, I replayed the situation in OCP 4.9.9 in PSI.

File: config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst (hunk 3 above)
Comment (avisiedo) [reply config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst]: Thanks @frasertweedale for your comments. I misread the systemd documentation and I wanted to set the values for 30 max intents to restart and a delay of 10 seconds between the retries (my thoughts).

I like "10 times in 5 minutes". I will change it. I want to limit this to avoid the systemd.unit is restarting the systemd.unit if for some reason the command fails continuously. About RestartSec I will set this to 10 seconds, so that if the process fails, we don't have to await too much until the RPM package is tried to be installed again.

I have seen in PSI cluster that first time is always failing. I am trying to retrieve a traceback that I saw the last Friday, so it could point out directly to the root cause.

In short: If the systemd.unit is failing continuously, it will be restarted with a delay of 10 seconds from the last failure; moreover the systemd.unit will stop restarting if the threshold is reached for 10 restarts in 5 minutes, avoiding that this is restarting and downloading the RPM packages continuously.

File: config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst (hunk 4 above)
Comment (avisiedo) [reply config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst]: Thanks, removing and keeping it cleaner.

File: config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst (hunk 3 above)
Comment (frasertweedale) [reply config/static/nodes/rpm-overrides/install-custom-rpm.yaml.envsubst]: Thanks, the new values LGTM.   I think we can do a final review during collab review session today, then merge.

File: config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst (hunk 2 above)
Comment (tiran) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: ``network-online`` does not meant that you can make network connections and lookup hostnames in DNS. You have to wait until after name lookups work:

```
After=network-online.target nss-lookup.target
Wants=network-online.target
```

Comment (tiran) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: https://www.freedesktop.org/software/systemd/man/systemd.special.html

> A target that should be used as synchronization point for all host/network name service lookups.

Comment (avisiedo) [reply config/static/nodes/cri-o/install-custom-cri-o.yaml.envsubst]: thanks @tiran for the comments and the link to the documentation; updating the systemd unit generated.

File: config/static/nodes/cgroup-v2/kustomization.yaml [hunk 5]
```diff
@@ -2,5 +2,3 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
```
Comment (avisiedo) [inline config/static/nodes/cgroup-v2/kustomization.yaml]: - Remove this file from the repository.
- Add this path to the .gitignore contents.
- Generate the file from the Makefile rule.

File: config/static/nodes/cgroup-v2/Makefile [hunk 6]
```diff
@@ -0,0 +1,6 @@
+POOL ?= worker master
+
+.PHONY: configure
+configure:
+	@git checkout kustomization.yaml
```
Comment (avisiedo) [inline config/static/nodes/cgroup-v2/Makefile]: - Remove checkout, and generate the head of the file.

File: config/static/nodes/rpm-overrides/Makefile [hunk 7]
```diff
@@ -0,0 +1,26 @@
+POOL ?= worker master
+RPM_PACKAGES ?=
+# RPM_RUNC ?= https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm
+# RPM_RUNC_PKG_NAME ?= runc-1.0.0-990.rhaos4.8.gitcd80260.el8
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/Makefile]: Cleanup commented lines

File: config/static/nodes/rpm-overrides/Makefile [hunk 8]
```diff
@@ -0,0 +1,26 @@
+POOL ?= worker master
+RPM_PACKAGES ?=
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/Makefile]: Add description for the variables

File: config/static/nodes/rpm-overrides/Makefile [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+# RPM_RUNC ?= https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm
+# RPM_RUNC_PKG_NAME ?= runc-1.0.0-990.rhaos4.8.gitcd80260.el8
+
+.PHONY: configure
+configure:
+ifneq (,$(RPM_PACKAGES))
+	git checkout kustomization.yaml
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/Makefile]: The same as before; do not checkout this file, and generate the head from here, and delete kustomize.yaml from repository and don't checkout it.

File: config/static/nodes/rpm-overrides/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# Setting the runc environment
+
+This configuration install into the pool indicated a custom runc version by
+installing a RPM package. By default it installs the package for "worker" and
+"master" nodes, but we can change that behavior.
+
+Basic usage:
+
+```shell
+export RPM_PACKAGES="https://..."
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/README.md]: Add description an example about what is expected on this variable.

File: config/static/nodes/rpm-overrides/README.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+make -C config/static/nodes/runc configure
+kustomize build config/static/nodes/runc | oc create -f -
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/README.md]: Wrong reference into the paths, it should be: ` config/static/nodes/rpm-overrides`.

File: config/static/nodes/rpm-overrides/README.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+For installing only in one specific pool:
+
+```shell
+export RPM_PACKAGES="https://..."
+export POOL="master"
+make -C config/static/nodes/runc configure
+kustomize build config/static/nodes/runc | oc create -f -
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/README.md]: Wrong reference into the paths, it should be: `config/static/nodes/rpm-overrides`.

File: config/static/nodes/rpm-overrides/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+>
+> It is recommended to use `?=' operator to allow overriding the values form
+> the CLI when invoking make configure
+
+## Installing a custom rpm package for runc
+
+The below install a custom runc package into all the nodes.
+
+```shell
+POOL="master worker" RPM_PACKAGES=https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm make configure
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/README.md]:
```suggestion
POOL="master worker" RPM_PACKAGES="https://ftweedal.fedorapeople.org/runc-1.0.0-990.rhaos4.8.gitcd80260.el8.x86_64.rpm" make configure
```

File: config/static/nodes/rpm-overrides/kustomization.yaml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (avisiedo) [inline config/static/nodes/rpm-overrides/kustomization.yaml]: Generate the head of this file into the Makefile, remove this file from the repository, and add this file to .gitignore file.

File: config/static/nodes/userns-subid/Makefile [hunk 15]
```diff
@@ -0,0 +1,17 @@
+CRIO_USERNS_CONF_BASE64:=$(shell cat files/99-crio-userns.conf | base64 --wrap=0)
+SUBUID_BASE64:=$(shell cat files/subuid | base64 --wrap=0)
+SUBGID_BASE64:=$(shell cat files/subgid | base64 --wrap=0)
+POOL ?= worker master
+
+.PHONY: configure
+configure:
+	git checkout kustomization.yaml
```
Comment (avisiedo) [inline config/static/nodes/userns-subid/Makefile]: Remove this checkout operation.

File: config/static/nodes/userns-subid/kustomization.yaml [hunk 16]
```diff
@@ -0,0 +1,4 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (avisiedo) [inline config/static/nodes/userns-subid/kustomization.yaml]: The same as the previous ones.
